Shared/UserAgent.swift

```diff
@@ -35,33 +35,113 @@ open class UserAgent {
     }
-
-    public static func oppositeUserAgent() -> String {
-        let isDefaultUADesktop = UserAgent.isDesktop(ua: UserAgent.defaultUserAgent())
+  
+    public static func oppositeUserAgent(domain: String) -> String {
+        let isDefaultUADesktop = UserAgent.isDesktop(ua: UserAgent.getUserAgent(domain: domain))
         if isDefaultUADesktop {
-            return mobileUserAgent()
+            return UserAgent.getUserAgent(domain: domain, platform: .Mobile)
+        } else {
+            return UserAgent.getUserAgent(domain: domain, platform: .Desktop)
+        }
+    }
+    
+    public static func getUserAgent(domain: String, platform: UserAgentPlatform) -> String {
+        switch platform {
+        case .Desktop:
+            if let customUA = CustomUserAgentConstant.desktopUserAgent[domain] {
+                return customUA
+            } else {
+                return desktopUserAgent()
+            }
+        case .Mobile:
+            if let customUA = CustomUserAgentConstant.mobileUserAgent[domain] {
+                return customUA
+            } else {
+                return mobileUserAgent()
+            }
+        }
+    }
+    
+    public static func getUserAgent(domain: String = "") -> String {
+        // As of iOS 13 using a hidden webview method does not return the correct UA on
+        // iPad (it returns mobile UA). We should consider that method no longer reliable.
+        if UIDevice.current.userInterfaceIdiom == .pad {
+            return getUserAgent(domain: domain, platform: .Desktop)
         } else {
-            return desktopUserAgent()
+            return getUserAgent(domain: domain, platform: .Mobile)
         }
     }
 }
+
+public enum UserAgentPlatform {
+    case Desktop
+    case Mobile
+}
+
+public struct CustomUserAgentConstant {
+    public static let mobileUserAgent = [
+        "whatsapp.com":
+        UserAgentBuilder.defaultMobileUserAgent().userAgent(),
+        "paypal.com":
+        UserAgentBuilder.defaultMobileUserAgent().userAgent(),
+        "yahoo.com":
+        UserAgentBuilder.defaultMobileUserAgent().userAgent()]
+    public static let desktopUserAgent = [
+        "whatsapp.com":
+        UserAgentBuilder.defaultDesktopUserAgent().clone(extensions: "Version/\(AppInfo.appVersion) \(UserAgent.uaBitSafari)"),
+        "paypal.com":
+        UserAgentBuilder.defaultDesktopUserAgent().clone(extensions: "Version/\(AppInfo.appVersion) \(UserAgent.uaBitSafari)"),
+        "yahoo.com":
+        UserAgentBuilder.defaultDesktopUserAgent().clone(extensions: "Version/\(AppInfo.appVersion) \(UserAgent.uaBitSafari)")]
+}
+
+public struct UserAgentBuilder {
+    // User agent components
+    fileprivate var product = ""
+    fileprivate var systemInfo = ""
+    fileprivate var platform = ""
+    fileprivate var platformDetails = ""
+    fileprivate var extensions = ""
+    
+    init(product: String, systemInfo: String, platform: String, platformDetails: String, extensions: String) {
```

Comment:
nit: product default could be "Mozilla/5.0", platform default could be "AppleWebKit/605.1.15", platformDetails default could be "(KHTML, like Gecko)"
or make these static const in the class, because they are re-used strings.

Reply:
Moved to the class above and made them static so they can be changed once.